core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -282,6 +317,188 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
+    val taskSet = FakeTask.createTaskSet(numTasks = 500, stageId = 0, stageAttemptId = 0)
+    taskScheduler.submitTasks(taskSet)
+
+    val offers = IndexedSeq(
+      new WorkerOffer("executor1", "host1", 1),
+      new WorkerOffer("executor2", "host1", 1),
+      new WorkerOffer("executor3", "host2", 1)
+    )
+    // We should check the node & exec blacklists, but only O(numOffers), not O(numPendingTasks)
+    // times.  Its O(numOffers), not exactly numOffers, because we offer resources multiple
+    // times.  An upper bound on the worst case is -- we check it once for every core at each
+    // locality level.  (We could tighten the bound a bit but that should be a good enough check.)
+    val numCoresOnAllOffers = offers.map(_.cores).sum
+    val numLocalityLevels = TaskLocality.values.size
+    val maxBlacklistChecks = numCoresOnAllOffers * numLocalityLevels
+
+    // Setup the blacklist, and get back a list of the executors & nodes that have any blacklisting
+    // (even implicit blacklisting).
+
+    nodeBlacklist.foreach { node =>
+      when(stageToMockTaskSetBlacklist(0).isNodeBlacklistedForTaskSet(node)).thenReturn(true)
+    }
+    execBlacklist.foreach { exec =>
+      when(stageToMockTaskSetBlacklist(0).isExecutorBlacklistedForTaskSet(exec)).thenReturn(true)
+    }
+    val nodesForBlacklistedExecutors = offers.filter { offer =>
+      execBlacklist.contains(offer.executorId)
+    }.map(_.host).toSet.toSeq
+    val nodesToCheck = nodeBlacklist ++ nodesForBlacklistedExecutors
+    val execsForBlacklistedNodes = offers.filter { offer =>
+      nodeBlacklist.contains(offer.host)
+    }.map(_.executorId).toSeq
+    val executorsToCheck = execBlacklist ++ execsForBlacklistedNodes
+
+    // Schedule a taskset, do a bit of basic sanity checking (that our test is operating the way its
+    // supposed to).
+    val firstTaskAttempts = taskScheduler.resourceOffers(offers).flatten
+    assert(firstTaskAttempts.size === 3 - executorsToCheck.size)
+    assert(firstTaskAttempts.size < offers.size)
+
+    // Now check that we haven't made too many calls to any of the blacklist methods.
+    verify(stageToMockTaskSetBlacklist(0), atMost(maxBlacklistChecks))
+      .isNodeBlacklistedForTaskSet(anyString())
+    nodesToCheck.foreach { node =>
+      verify(stageToMockTaskSetBlacklist(0), never)
+        .isNodeBlacklistedForTask(meq(node), anyInt())
+    }
+    executorsToCheck.foreach { exec =>
+      // If we had a node-blacklist, then we could tighten the next check to *never*.  But, it also
+      // doesn't particular matter if we the executor check happens in addition.
+      verify(stageToMockTaskSetBlacklist(0), atMost(maxBlacklistChecks))
+        .isExecutorBlacklistedForTaskSet(exec)
+      verify(stageToMockTaskSetBlacklist(0), never)
+        .isExecutorBlacklistedForTask(meq(exec), anyInt())
+    }
+  }
+
+  test("(node, taskset) blacklist prevents per-task checks") {
```

Comment:
How about "Blacklisted node for entire task set prevents per-task blacklist checks"? (found the existing version non-obvious in what it meant). And for below maybe "Blacklisted executor for entire task set prevents per-task blacklist checks"?